client/client_test.go

```diff
@@ -131,7 +131,7 @@ func TestIntegration(t *testing.T) {
 }
 
 func newContainerd(cdAddress string) (*containerd.Client, error) {
-	return containerd.New(cdAddress, containerd.WithTimeout(60*time.Second), containerd.WithDefaultRuntime("io.containerd.runtime.v1.linux"))
+	return containerd.New(cdAddress, containerd.WithTimeout(60*time.Second))
```

Comment:
This was needed in linux to support older containerd daemons that are not EOL yet.

Reply:
That's odd, because it was only added when containerd 1.3 was vendor'd. Previous containerd has this as the default value already, from my reading, and containerd changed the default to the runtime v2 shim.

Reply:
@TBBle: When you get back to this, see if the problem is that the default is being controlled by the vendored part, not the daemon.

Reply:
Containerd client(vendored) unfortunately can't do detection of the supported shims (although the client should always be backward compatible based on docs). So it will default to shim2 and fail on older daemons because they don't have any shim2 binary.

Reply:
Note that this isn't a problem for windows though as none of this works for old containerd in windows anyway.

Reply:
That's what I was afraid of. It seems that if the client sends unusable defaults, it's not really backwards-compatible; I'd think the daemon should be choosing the default runtime.
So I guess the best option is to make it an OS check and maybe a command-line switch, until we drop support for whatever version of containerd does work with the containerd client library's default runtime choice.

Reply:
That of course means it's not low-hanging fruit, so unless I find time to really hammer this out (including working out how to get the unit tests to catch this in future), I'll pull these two changes out of the PR.

Reply:
@tonistiigi : According to #1176, buildkit dropped support for containerd older than 1.3 already. Or am I misreading that?

Reply:
Yes and no. It is true #1176 depends on 1.3 but it still works with older daemons while being less efficient on releasing data in some cases. This is different than getting fatal "binary not found" error on every build.

Reply:
I've pulled this commit and the previous one (pulling the new containerd Linux V2 runtime into the travis build), I'll come back to this in a new pull request later.

I plan to introduce a command-line option to select containerd runtime, with legacy runc runtime as default on non-Windows (matching containerd 1.2), and hcsshim runtime as default on Windows (matching containerd 1.3).

I'll also try and reinstate the Containerd 1.2 unit tests in the Travis build, so I or someone else doesn't fall into this gap again.

Reply:
@tonistiigi Are you _sure_ that the current master works with containerd 1.2? I tried enabling the containerd 1.2 unit tests, and a fair few of them died trying to call `AddResource` against containerd, which was added in 1.3.0.

It looks like anything hitting `func (s *sourceOp) Exec` for the `containerimage` source will end up in `func (cm *cacheManager) GetByBlob`, which calls `LeaseManager.AddResource` and hence fails against containerd 1.2.

It's very possible I'm overlooking something obvious, but I do not currently see any path that can bypass those calls when the API is not present.

Is there just a class of tests/features that are not valid with containerd 1.2?